### PR TITLE
Post research changes

### PR DESCRIPTION
**Pull request by jonathanglassman**
- Expanded method code examples
- Added internal links to arguments in each section
- Moved arguments to under the method code
- Added "Back to top" links
- Renamed API reference to Error codes
- Added information on statuses at top of Get Message Status section
- Tidied up some examples
- Added questions

### CODE REVIEW

**Comment by quis** [comment]
Haven’t had a chance to review the whole thing, but a couple of things I think would help:

# Only include required arguments in example

For example, users are less likely to be confused if they copy and paste this:
```python
response = notifications_client.send_sms_notification(
   phone_number='07900900123',
   template_id='f33517ff-2a88-4f6e-b855-c550268ce08a',
)
```

…instead of this:

```python
response = notifications_client.send_sms_notification(
   phone_number='+447900900123', # required string
   template_id='f33517ff-2a88-4f6e-b855-c550268ce08a', # required UUID string
   personalisation={
       'KEY': 'VALUE',
       'KEY': 'VALUE',
       ...
       }, # optional dict - specifies template parameters
   reference='STRING', # optional string - identifies notification(s)
   sms_sender_id='8e222534-7f05-4972-86e3-17c5d9f894e2' # optional UUID string
)
```

alternatively, you could specify `None` for the optional arguments, like the current documentation:

image

My preference is for the first example though.

# Label each argument as required/optional

Currently the outline of the documentation looks like:
```markdown
- Arguments
  - Required
    - Argument name
    - Argument name
  - Optional
    - Argument name
    - Argument name
    _ Argument name
```

I think it would work better this like (because users tend to very quickly narrow in on the one thing they care about):

```markdown
- Arguments
  - Argument name (required)
  - Argument name (required)
  - Argument name (optional)
  - Argument name (optional)
  - Argument name (optional)
```

**Comment by quis** [comment]
# Back to top links

Not convinced by the utility of these.

# Error messages/codes 

We should think more carefully about which ones we show where. Currently, for example, the first error documented under _Send a text message_ is `429: Rate limit exceeded`. This is unlikely to be an error that someone just integrating with Notify for the first time 

I think that errors specific to rate limits should go in a rate limit section of the doc. Only errors specific to sending messages (eg missing placeholders) should go with the sending message bit. 

I also think we should consider which error messages need documenting (probably those that a developer should be considering catching) and those that are self-explanatory (eg trial mode).

File: Notify-Tech-Docs/source/documentation/python.md [hunk 1]
```diff
@@ -69,338 +109,304 @@ If you are using the [test API key](/#test), all your messages will come back as
...
-
-## Send a letter
-
-When your service first signs up to GOV.UK Notify, you’ll start in trial mode. You can only send letters in live mode.
-
-### Method
-
-1. Add the following method code to your application:
+```python
+reference=’STRING’ # optional string - identifies notification(s)
```
Comment (servingUpAces) [inline Notify-Tech-Docs/source/documentation/python.md]: the quotes should be ' not ’

**Review by servingUpAces — COMMENTED** [review]
This is ok for me.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 2]
```diff
@@ -69,338 +109,304 @@ If you are using the [test API key](/#test), all your messages will come back as
...
+#### email_address (required)
 
-##### Optional
+The email address of the recipient, only required for email notifications.
 
-`reference`
+#### template_id (required)
 
-An unique identifier for the notification. This reference can identify a single unique notification or a batch of multiple notifications.
+The ID of the template. You can find this by logging into GOV.UK Notify and going to the _Templates_ page.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: _signing into_ not _logging into_

File: Notify-Tech-Docs/source/documentation/python.md [hunk 3]
```diff
@@ -412,42 +418,86 @@ This API call returns the status of all messages. You can either get the status
...
+```python
+reference=’STRING’ # optional string - identifies notification(s)
+```
+
+#### older_than (optional)
+
+Input the ID of a notification into this argument. If you use this argument, the next 250 received notifications older than the given ID are returned.
+
+```python
+older_than=’740e5834-3a29-46b4-9a6f-16142fde533a’ # optional string - notification ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Quotes should be straight (`'` not `‘`)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 4]
```diff
@@ -412,42 +418,86 @@ This API call returns the status of all messages. You can either get the status
...
+|`technical-failure`|Email / Text: Notify had a technical failure; you can try to send the message again. <br><br>Letter: Notify had an unexpected error while sending to our printing provider. <br><br>You can omit this argument to ignore this filter.|Yes|Yes||
+|`accepted`|Notify is printing and posting the letter|||Yes|
 
-Run your application to send a request to API Notification client. It will return the next oldest messages from the specified notification ID.
+#### reference (optional)
+
+A unique identifier. This reference can identify a single unique notification or a batch of multiple notifications.
+
+```python
+reference=’STRING’ # optional string - identifies notification(s)
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Quotes should be straight (`'` not `‘`)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 5]
```diff
@@ -553,105 +568,95 @@ If this argument is omitted, the most recent 250 notifications are returned.
...
+  `f33517ff-2a88-4f6e-b855-c550268ce08a` # required string - template ID
 )
 ```
 
-Run your application to send a request to API Notification client.
+### Arguments
+
+#### template_id (required)
+
+The ID of the template. You can find this by logging into GOV.UK Notify and going to the _Templates_ page.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: _Signing into_ not _logging into_.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 6]
```diff
@@ -553,105 +568,95 @@ If this argument is omitted, the most recent 250 notifications are returned.
...
 
 This will return the latest version of the template.
 
-Add the following method code to your application:
-
 ```python
 response = notifications_client.get_template_version(
-    'template_id',
-    `version` # integer required for version number
+    `f33517ff-2a88-4f6e-b855-c550268ce08a` # required string - template ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Straight quotes not backticks (`'` not `` ` ``)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 7]
```diff
@@ -553,105 +568,95 @@ If this argument is omitted, the most recent 250 notifications are returned.
...
+    ‘version’: NUMBER, # required string - template version
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Straight quotes not typographic quotes (`'` not `’`)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 8]
```diff
@@ -553,105 +568,95 @@ If this argument is omitted, the most recent 250 notifications are returned.
...
 {
-    "id": "template_id", # required
-    "type": "sms" | "email" | "letter", # required
-    "created_at": "created at", # required
-    "updated_at": "updated at", # required
-    "version": "version", # integer required
-    "created_by": "someone@example.com", # email required
-    "body": "Body of the notification", # required
-    "subject": "Subject of an email or letter notification, or None if an sms message"
+    "id": `f33517ff-2a88-4f6e-b855-c550268ce08a` # required string - template ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Straight quotes not backticks (`'` not `` ` ``)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 9]
```diff
@@ -678,17 +689,17 @@ If the request to the client is successful, you will receive a `dict` response.
...
         {
-            "id": "template_id", # required
-            "type": "sms" | "email" | "letter", # required
-            "created_at": "created at", # required
-            "updated_at": "updated at", # required
-            "version": "version", # integer required
-            "created_by": "someone@example.com", # email required
-            "body": "Body of the notification", # required
-            "subject": "Subject of an email or letter notification, or None if an sms message"
+            "id": `f33517ff-2a88-4f6e-b855-c550268ce08a` # required string - template ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Straight quotes not backticks (`'` not `` ` ``)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 10]
```diff
@@ -702,77 +713,67 @@ If no templates exist for a template type or there no templates for a service, y
...
 ```python
 {
-    "id": "notify_id", # required
-    "type": "sms" | "email" | "letter", # required
-    "version": "version", # integer required
-    "body": "Body of the notification", # required
-    "subject": "Subject of an email or letter notification, or None if an sms message"
+    "id": "740e5834-3a29-46b4-9a6f-16142fde533a", # required string - notification ID
+    "type": "sms / email / letter" , # required string
+    "version": NUMBER, # required string - template version
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Why has version changed from integer to string? Is this right?

File: Notify-Tech-Docs/source/documentation/python.md [hunk 11]
```diff
@@ -800,22 +797,29 @@ If the request to the client is successful, you will receive a `<generator objec
...
 ### Method
 
-Add the following method code to your application:
-
 ```python
 response = client.get_received_texts(older_than)
 ```
 
-You specify which texts to receive by inputting the ID of a received text message into the `older_than` argument.
+You can specify which texts to receive by inputting the ID of a received text message into the [`older_than`](/#get-one-page-of-received-text-messages-optional-arguments-older-than) argument.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: _text messages_ not _texts_

File: Notify-Tech-Docs/source/documentation/python.md [hunk 12]
```diff
@@ -800,22 +797,29 @@ If the request to the client is successful, you will receive a `<generator objec
...
 
-Run your application to send a request to API Notification client.
+### Arguments
+
+#### older_than (optional)
+
+Input the ID of a received text message into this argument. If you use this argument, the next 250 received text messages older than the given ID are returned.
+
+```python
+older_than=’740e5834-3a29-46b4-9a6f-16142fde533a’ # optional string - notification ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Straight quotes not typographic quotes (`'` not `’`)

File: Notify-Tech-Docs/source/documentation/python.md [hunk 13]
```diff
@@ -69,338 +109,304 @@ If you are using the [test API key](/#test), all your messages will come back as
...
-        reference=None,
-        email_reply_to_id=None
-    )
-    ```
+This is an email reply-to address specified by you to receive replies from your users. Your service cannot go live until at least one email address has been set up for this. To set up:
 
-1. Complete the required `email_address` and `template_id` arguments.
+1. Log into your GOV.UK Notify account.
+1. Go to _Settings_.
+1. Check that you are in the correct service. If you are not, click _Switch service_ in the top right corner of the screen and select the correct one.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Drop this step. We should assume that users are looking at the correct service.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 14]
```diff
@@ -69,338 +109,304 @@ If you are using the [test API key](/#test), all your messages will come back as
...
+```
 
-`template_id`
+### Arguments
 
-Find this by logging into GOV.UK Notify and going to the Templates page.
+#### email_address (required)
 
-##### Optional
+The email address of the recipient, only required for email notifications.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: The implication being that `email_address` is optional for non-email notifications? I’d lose this second clause, don’t think it makes sense or is necessary.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 15]
```diff
@@ -18,49 +20,87 @@ from notifications_python_client.notifications import NotificationsAPIClient
...
+reference='STRING', # optional string - identifies notification(s)
+```
+
+#### sms_sender_id (optional)
+
+A unique identifier of the sender of the text message notification. To set this up:
+
+1. Log into your GOV.UK Notify account.
+1. Go to _Settings_.
+1. Check that you are in the correct service. If you are not, click _Switch service_ in the top right corner of the screen and select the correct one.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Drop this step. We should assume that users are looking at the right service.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 16]
```diff
@@ -18,49 +20,87 @@ from notifications_python_client.notifications import NotificationsAPIClient
...
+### Arguments
+
+#### phone_number (required)
+
+The phone number of the recipient of the text message. This number can be UK or international.
+
+#### template_id (required)
 
-1. Complete the required `phone_number` and `template_id` arguments.
+The ID of the template. You can find this by logging into [GOV.UK Notify](https://www.notifications.service.gov.uk/) and going to the _Templates_ page.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: _signing into_ not _logging into_

File: Notify-Tech-Docs/source/documentation/python.md [hunk 17]
```diff
@@ -18,49 +20,87 @@ from notifications_python_client.notifications import NotificationsAPIClient
 notifications_client = NotificationsAPIClient(api_key)
 ```
 
-To get an API key, [log in to GOV.UK Notify](https://www.notifications.service.gov.uk/) and go to the _API integration_ page. Information on API keys can be found [here](/#api-keys).
+To get an API key, [log in to GOV.UK Notify](https://www.notifications.service.gov.uk/) and go to the _API integration_ page. More information can be found in the [API keys](/#api-keys) section.
 
 # Send a message
 
+GOV.UK Notify enables you to send text messages, emails and letters.
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Pretty sure _enables_ is against GOV.UK style. I’d go with something like ‘You can use GOV.UK Notify to send text messages, emails and letters.’

**Review by quis — CHANGES_REQUESTED** [review]
A bunch of small style changes, but otherwise think this is good to go.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 18]
```diff
@@ -384,7 +382,7 @@ If the request to the client is successful, you will receive the following `dict
   "type": "sms / letter / email", # required string
   "status": "sending / delivered / permanent-failure / temporary-failure / technical-failure", # required string
   "template": {
-    "Version": NUMBER # required string - template version
+    "Version": INTEGER
     "id": `f33517ff-2a88-4f6e-b855-c550268ce08a` # required string - template ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: More backticks `` ` `` – missed these first time, sorry.

File: Notify-Tech-Docs/source/documentation/python.md [hunk 19]
```diff
@@ -523,7 +521,7 @@ If the request to the client is successful, you will receive a `dict` response.
       "type": "sms / letter / email", # required string
       "status": "sending / delivered / permanent-failure / temporary-failure / technical-failure", # required string
       "template": {
-        "version": NUMBER # required string - template version
+        "version": INTEGER 
         "id": `f33517ff-2a88-4f6e-b855-c550268ce08a` # required string - template ID
```
Comment (quis) [inline Notify-Tech-Docs/source/documentation/python.md]: Backticks `` ` `` should be single quotes `'`

File: Notify-Tech-Docs/source/documentation/python.md (hunk 18 above)
Comment (jonathanglassman) [reply Notify-Tech-Docs/source/documentation/python.md]: Removed these and a few more

**Pull request merged**